tiledb/api/c_api/vfs/vfs_api.cc

```diff
@@ -0,0 +1,533 @@
+  ensure_vfs_is_valid(vfs);
+  throw_if_not_ok(vfs->remove_dir(tiledb::sm::URI(uri)));
+  return TILEDB_OK;
+}
+
+capi_return_t tiledb_vfs_is_file(
+    tiledb_vfs_t* vfs, const char* uri, int32_t* is_file) {
+  ensure_vfs_is_valid(vfs);
+  ensure_output_pointer_is_valid(is_file);
+
+  bool exists;
+  throw_if_not_ok(vfs->is_file(tiledb::sm::URI(uri), &exists));
+  *is_file = (int32_t)exists;
+
+  return TILEDB_OK;
+}
+
+capi_return_t tiledb_vfs_remove_file(tiledb_vfs_t* vfs, const char* uri) {
+  ensure_vfs_is_valid(vfs);
+  throw_if_not_ok(vfs->remove_file(tiledb::sm::URI(uri)));
+  return TILEDB_OK;
+}
+
+capi_return_t tiledb_vfs_dir_size(
+    tiledb_vfs_t* vfs, const char* uri, uint64_t* size) {
+  ensure_vfs_is_valid(vfs);
+  ensure_output_pointer_is_valid(size);
+  throw_if_not_ok(vfs->dir_size(tiledb::sm::URI(uri), size));
+  return TILEDB_OK;
+}
+
+capi_return_t tiledb_vfs_file_size(
+    tiledb_vfs_t* vfs, const char* uri, uint64_t* size) {
+  ensure_vfs_is_valid(vfs);
+  ensure_output_pointer_is_valid(size);
+  throw_if_not_ok(vfs->file_size(tiledb::sm::URI(uri), size));
+  return TILEDB_OK;
+}
+
+capi_return_t tiledb_vfs_move_file(
+    tiledb_vfs_t* vfs, const char* old_uri, const char* new_uri) {
+  ensure_vfs_is_valid(vfs);
+  throw_if_not_ok(
+      vfs->move_file(tiledb::sm::URI(old_uri), tiledb::sm::URI(new_uri)));
+  return TILEDB_OK;
+}
+
+capi_return_t tiledb_vfs_move_dir(
+    tiledb_vfs_t* vfs, const char* old_uri, const char* new_uri) {
+  ensure_vfs_is_valid(vfs);
+  throw_if_not_ok(
+      vfs->move_dir(tiledb::sm::URI(old_uri), tiledb::sm::URI(new_uri)));
+  return TILEDB_OK;
+}
+
+capi_return_t tiledb_vfs_copy_file(
+    tiledb_vfs_t* vfs, const char* old_uri, const char* new_uri) {
+  ensure_vfs_is_valid(vfs);
+  throw_if_not_ok(
+      vfs->copy_file(tiledb::sm::URI(old_uri), tiledb::sm::URI(new_uri)));
+  return TILEDB_OK;
+}
+
+capi_return_t tiledb_vfs_copy_dir(
+    tiledb_vfs_t* vfs, const char* old_uri, const char* new_uri) {
+  ensure_vfs_is_valid(vfs);
+  throw_if_not_ok(
+      vfs->copy_dir(tiledb::sm::URI(old_uri), tiledb::sm::URI(new_uri)));
+  return TILEDB_OK;
+}
+
+capi_return_t tiledb_vfs_open(
+    tiledb_vfs_t* vfs,
+    const char* uri,
+    tiledb_vfs_mode_t mode,
+    tiledb_vfs_fh_t** fh) {
+  ensure_vfs_is_valid(vfs);
+  ensure_output_pointer_is_valid(fh);
+
+  // Create VFS file handle
+  auto fh_uri = tiledb::sm::URI(uri);
+  if (fh_uri.is_invalid()) {
+    throw CAPIStatusException(std::string("Invalid TileDB object: ") + "uri");
+  }
+  auto vfs_mode = static_cast<tiledb::sm::VFSMode>(mode);
+  *fh = tiledb_vfs_fh_t::make_handle(fh_uri, vfs->vfs(), vfs_mode);
+
+  // Open VFS file
+  auto st{(*fh)->open()};
+  if (!st.ok()) {
+    tiledb_vfs_fh_t::break_handle(*fh);
+    return TILEDB_ERR;
+  }
+
+  return TILEDB_OK;
+}
+
+capi_return_t tiledb_vfs_close(tiledb_vfs_fh_t* fh) {
+  ensure_output_pointer_is_valid(fh);
+  ensure_vfs_fh_is_valid(fh);
+  throw_if_not_ok(fh->close());
```

Comment:
Do we need a `tiledb_vfs_fh_t::break_handle(fh)` here?

Reply:
Updated: I believe the handle should _only_ be broken in the `free` function. After close, there can still be a valid `vfs_fh` object until it's freed. IMO this should be changed but that's how the code currently stands